Make it possible to compile in Msys2/ucrt64

hly1204 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository.
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style).
- [ ] I've documented every `RZ_API` function and struct this PR changes.
- [ ] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the [Rizin book](https://github.com/rizinorg/book) with the relevant information (if needed).
- [ ] I've used AI tools to generate fully or partially these code changes and I'm sure the changes are not copyrighted by somebody else.

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request.
If you have used AI tools to generate these code changes, please disclose software used, model name. -->

It is a patch to make it possible to compile in [MSYS2/UCRT64](https://www.msys2.org/docs/environments/). I donot know whether the community want to use MSYS2 for the project or not, and I am not sure that what I have modified will break the building for exist platforms. I appreciate if one would left some comments.

The building steps in MSYS2/UCRT64 are as follows:

```shell
pacman -S mingw-w64-ucrt-x86_64-toolchain \
          mingw-w64-ucrt-x86_64-python \
          mingw-w64-ucrt-x86_64-meson \
          mingw-w64-ucrt-x86_64-capstone \
          mingw-w64-ucrt-x86_64-pcre2 \
          mingw-w64-ucrt-x86_64-zydis \
          mingw-w64-ucrt-x86_64-libmspack \
          mingw-w64-ucrt-x86_64-xxhash \
          mingw-w64-ucrt-x86_64-libblake3 \
          mingw-w64-ucrt-x86_64-openssl \
          mingw-w64-ucrt-x86_64-libtree-sitter \
          mingw-w64-ucrt-x86_64-zlib \
          mingw-w64-ucrt-x86_64-lz4 \
          mingw-w64-ucrt-x86_64-libzip \
          mingw-w64-ucrt-x86_64-zstd

meson setup build \
    -D use_sys_capstone=enabled \
    -D use_sys_pcre2=enabled \
    -D use_sys_zydis=enabled \
    -D use_sys_libmspack=enabled \
    -D use_sys_xxhash=enabled \
    -D use_sys_blake3=enabled \
    -D use_sys_openssl=enabled \
    -D use_sys_tree_sitter=enabled \
    -D use_sys_zlib=enabled \
    -D use_sys_lz4=enabled \
    -D use_sys_libzip=enabled \
    -D use_sys_libzip_openssl=true \
    -D use_sys_libzstd=enabled
cd build
ninja
```

hly1204 committed:
Make it possible to compile in Msys2/ucrt64

notxvilka reviewed (commented):
See also https://github.com/rizinorg/rizin/issues/337

Also see all red CI jobs caused by these changes.

hly1204 commented:
> See also #337
> 
> Also see all red CI jobs caused by these changes.

Thank you! I will try to fix it.

hly1204 committed:
Try fix *nix build

hly1204 commented:
It seems that the def. of `execv` is `int execv(const char *path, char *const argv[]);` (<https://linux.die.net/man/3/execv>), while the def. of `_execv` is `intptr_t _execv(const char *cmdname, const char *const *argv);` (<https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/execv-wexecv?view=msvc-170>). The later needs an additional `const` qualifier. A work-around is to use the following macro, but it seems lost the auto type check. I donot know if this is a acceptable solution.

```c
#if _MSC_VER || __MINGW32__
/* https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/execv-wexecv?view=msvc-170 */
#define rz_sys_execv(cmdname, argv) _execv((cmdname), (const char *const *)(argv))
#else
#define rz_sys_execv(path, argv) execv((path), (char *const *)(argv))
#endif
```

notxvilka commented:
> It seems that the def. of `execv` is `int execv(const char *path, char *const argv[]);` (https://linux.die.net/man/3/execv), while the def. of `_execv` is `intptr_t _execv(const char *cmdname, const char *const *argv);` (https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/execv-wexecv?view=msvc-170). The later needs an additional `const` qualifier. A work-around is to use the following macro, but it seems lost the auto type check. I donot know if this is a acceptable solution.
> 
> ```c
> #if _MSC_VER || __MINGW32__
> /* https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/execv-wexecv?view=msvc-170 */
> #define rz_sys_execv(cmdname, argv) _execv((cmdname), (const char *const *)(argv))
> #else
> #define rz_sys_execv(path, argv) execv((path), (char *const *)(argv))
> #endif
> ```

This kind of thing is better to check on the Meson level (just see how other functions are checked, for example `pipe2` function in the top `meson.build` file, along with some other functions in this code. And then picked up in the userconf header.

cc @well-mannered-goat for the debug changes

well-mannered-goat reviewed (commented):
I would recommend keeping a different file for Msys2/ucrt64 compilation although windows related, your changes break the appveyor CI due to compilation difference.
```
FAILED: librz/debug/rz_debug.lib.p/p_native_windows_windows_debug.c.obj 
"cl" "-Ilibrz\debug\rz_debug.lib.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-I..\librz\bin\format\elf" "-I..\librz\bin\format\dmp" "-I..\librz\bin\format\mdmp" "-I..\librz\bin\format\pe" "-I..\subprojects\rzgdb\include" "-I..\subprojects\rzgdb\include\gdbclient" "-I..\subprojects\rzgdb\include\gdbserver" "-Isubprojects\rzwinkd" "-I..\subprojects\rzwinkd" "-Ilibrz\util\sdb\src" "-I..\librz\util\sdb\src" "-I..\librz\bin\format" "-I..\librz\arch\isa" "-I..\librz\arch\isa_gnu" "-Ilibrz\arch" "-I..\librz\arch" "-I..\librz\type\parser" "-I..\subprojects\rzqnx\include" "-I..\subprojects\rzw32dbg_wrap\include" "/MT" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "/Gw" "-DPSAPI_VERSION=1" "-DRZ_PLUGIN_INCORE=1" "-DSUPPORTS_PCRE2_JIT" "-DZYDIS_STATIC_BUILD" "/Fdlibrz\debug\rz_debug.lib.p\p_native_windows_windows_debug.c.pdb" /Folibrz/debug/rz_debug.lib.p/p_native_windows_windows_debug.c.obj "/c" ../librz/debug/p/native/windows/windows_debug.c
../librz/debug/p/native/windows/windows_debug.c(1497): error C2065: 'POBJECT_TYPE_INFORMATION': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1497): error C2146: syntax error: missing ';' before identifier 'objectTypeInfo'
../librz/debug/p/native/windows/windows_debug.c(1497): error C2065: 'objectTypeInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1497): warning C4047: '=': 'int' differs in levels of indirection from 'void *'
../librz/debug/p/native/windows/windows_debug.c(1498): error C2065: 'objectTypeInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1504): error C2065: 'PSYSTEM_HANDLE_INFORMATION': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1504): error C2146: syntax error: missing ';' before identifier 'handleInfo'
../librz/debug/p/native/windows/windows_debug.c(1504): error C2065: 'handleInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1504): warning C4047: '=': 'int' differs in levels of indirection from 'void *'
../librz/debug/p/native/windows/windows_debug.c(1510): error C2065: 'handleInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1510): error C2065: 'PSYSTEM_HANDLE_INFORMATION': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1510): error C2146: syntax error: missing ';' before identifier 'malloc'
../librz/debug/p/native/windows/windows_debug.c(1511): error C2065: 'handleInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1511): warning C4022: 'w32_NtQuerySystemInformation': pointer mismatch for actual parameter 2
../librz/debug/p/native/windows/windows_debug.c(1513): error C2065: 'handleInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1513): warning C4022: 'realloc': pointer mismatch for actual parameter 1
../librz/debug/p/native/windows/windows_debug.c(1515): error C2065: 'handleInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1515): error C2065: 'PSYSTEM_HANDLE_INFORMATION': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1515): error C2146: syntax error: missing ';' before identifier 'tmp'
../librz/debug/p/native/windows/windows_debug.c(1530): error C2065: 'handleInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1530): error C2223: left of '->Count' must point to struct/union
../librz/debug/p/native/windows/windows_debug.c(1531): error C2065: 'SYSTEM_HANDLE_ENTRY': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1531): error C2146: syntax error: missing ';' before identifier 'handle'
../librz/debug/p/native/windows/windows_debug.c(1531): error C2065: 'handle': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1531): error C2065: 'handleInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1531): error C2223: left of '->Handle' must point to struct/union
../librz/debug/p/native/windows/windows_debug.c(1535): error C2065: 'handle': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1535): error C2224: left of '.OwnerPid' must have struct/union type
../librz/debug/p/native/windows/windows_debug.c(1538): error C2065: 'handle': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1538): error C2224: left of '.HandleValue' must have struct/union type
../librz/debug/p/native/windows/windows_debug.c(1538): error C2198: 'w32_NtDuplicateObject': too few arguments for call
../librz/debug/p/native/windows/windows_debug.c(1541): error C2065: 'objectTypeInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1541): warning C4022: 'w32_NtQueryObject': pointer mismatch for actual parameter 3
../librz/debug/p/native/windows/windows_debug.c(1545): error C2065: 'objectTypeInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1545): error C2223: left of '->TypeName' must point to struct/union
../librz/debug/p/native/windows/windows_debug.c(1545): error C2168: 'wcscmp': too few actual parameters for intrinsic function
../librz/debug/p/native/windows/windows_debug.c(1549): error C2065: 'objectTypeInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1549): error C2223: left of '->GenericMapping' must point to struct/union
../librz/debug/p/native/windows/windows_debug.c(1550): error C2065: 'handle': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1550): error C2224: left of '.AccessMask' must have struct/union type
../librz/debug/p/native/windows/windows_debug.c(1553): error C2065: 'handle': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1553): error C2224: left of '.AccessMask' must have struct/union type
../librz/debug/p/native/windows/windows_debug.c(1556): error C2065: 'handle': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1556): error C2224: left of '.AccessMask' must have struct/union type
../librz/debug/p/native/windows/windows_debug.c(1573): error C2065: 'handle': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1573): error C2224: left of '.HandleValue' must have struct/union type
../librz/debug/p/native/windows/windows_debug.c(1573): warning C4047: 'function': 'int' differs in levels of indirection from 'char *'
../librz/debug/p/native/windows/windows_debug.c(1573): warning C4024: 'rz_debug_desc_new': different types for formal and actual parameter 1
../librz/debug/p/native/windows/windows_debug.c(1573): warning C4047: 'function': 'char *' differs in levels of indirection from 'int'
../librz/debug/p/native/windows/windows_debug.c(1573): warning C4024: 'rz_debug_desc_new': different types for formal and actual parameter 2
../librz/debug/p/native/windows/windows_debug.c(1573): error C2198: 'rz_debug_desc_new': too few arguments for call
../librz/debug/p/native/windows/windows_debug.c(1581): error C2065: 'objectTypeInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1581): error C2223: left of '->TypeName' must point to struct/union
../librz/debug/p/native/windows/windows_debug.c(1581): error C2198: 'rz_utf16_to_utf8_l': too few arguments for call
../librz/debug/p/native/windows/windows_debug.c(1582): error C2065: 'handle': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1582): error C2224: left of '.HandleValue' must have struct/union type
../librz/debug/p/native/windows/windows_debug.c(1582): warning C4047: 'function': 'int' differs in levels of indirection from 'char *'
../librz/debug/p/native/windows/windows_debug.c(1582): warning C4024: 'rz_debug_desc_new': different types for formal and actual parameter 1
../librz/debug/p/native/windows/windows_debug.c(1582): warning C4047: 'function': 'char *' differs in levels of indirection from 'int'
../librz/debug/p/native/windows/windows_debug.c(1582): warning C4024: 'rz_debug_desc_new': different types for formal and actual parameter 2
../librz/debug/p/native/windows/windows_debug.c(1582): error C2198: 'rz_debug_desc_new': too few arguments for call
../librz/debug/p/native/windows/windows_debug.c(1594): error C2065: 'objectTypeInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1594): warning C4022: 'free': pointer mismatch for actual parameter 1
../librz/debug/p/native/windows/windows_debug.c(1595): error C2065: 'handleInfo': undeclared identifier
../librz/debug/p/native/windows/windows_debug.c(1595): warning C4022: 'free': pointer mismatch for actual parameter 1
[1976/2406] Compiling C object librz/debug/rz_debug.lib.p/p_native_maps_windows_maps.c.obj
[1977/2406] Compiling C object librz/debug/rz_debug.lib.p/p_native_windows_windows_message.c.obj
ninja: build stopped: subcommand failed.
```
https://ci.appveyor.com/project/rizinorg/rizin/builds/53013061/job/kj2gcflhtoh4u742

hly1204 commented:
Thanks and sorry for the late reply. I will check the build for MSVC. I have removed 2 definitions since there are duplicate ones in MinGW64. But it seems MSVC still needs these definitions (from my point of view, MinGW64 should have the exact same naming structs with MSVC), I will try to figure it out.

upd: Microsoft actully didnot provide associated header file (see: <https://learn.microsoft.com/en-us/windows/win32/api/winternl/nf-winternl-ntqueryobject#remarks>), and the definition of these `struct`s in MinGW64 is (mostly, I have checked the different field and all fields defined after it are not used, we only use `TypeName` and `GenericMapping` of `struct _OBJECT_TYPE_INFORMATION`, and we use a FIXED size (0x1000 = 4096) to allocate, so I think the behavior is not changed) identical to the ones I have removed in the memory layout, only the names of variables are different. Since this file in MinGW64 is in public domain, I think we could change the definition to MinGW64's (only a few changes are required).

hly1204 committed:
Try fix windows & *nix build

hly1204 committed:
Merge branch 'dev' into mingw

hly1204 commented:
Sorry I will fix the missing definition of `struct _SYSTEM_HANDLE_ENTRY`,

```c
typedef struct _SYSTEM_HANDLE_ENTRY {
	ULONG OwnerPid;
	BYTE ObjectType;
	BYTE HandleFlags;
	USHORT HandleValue;
	PVOID ObjectPointer;
	ULONG AccessMask;
} SYSTEM_HANDLE_ENTRY, *PSYSTEM_HANDLE_ENTRY;
```

the previous one is `struct _SYSTEM_HANDLE`

```c
typedef struct _SYSTEM_HANDLE {
	ULONG ProcessId;
	BYTE ObjectTypeNumber;
	BYTE Flags;
	USHORT Handle;
	PVOID Object;
	ACCESS_MASK GrantedAccess;
} SYSTEM_HANDLE, *PSYSTEM_HANDLE;
```

Their memory layouts are identical (ULONG = unsigned long, ACCESS_MASK = DWORD = unsigned long).

hly1204 committed:
Fix missing definition of `_SYSTEM_HANDLE_ENTRY`
